### PR TITLE
Yf make cram valid input

### PR DESCRIPTION
**Pull request by yfarjoun**
### Description

PR that makes CRAM file valid input (sets up referenceSequence for the reader, and updates documentation)
Also includes contributions from #683 (thanks @ekazachkova!! 🥇)

_Give your PR a **concise** yet **descriptive** title_
_Please explain the changes you made here._
_Explain the **motivation** for making this change. What existing problem does the pull request solve?_
_Mention any issues fixed, addressed or otherwise related to this pull request, including issue numbers or hard links for issues in other repos._
_You can delete these instructions once you have written your PR description._

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [ ] Added or modified tests to cover changes and any new functionality
- [ ] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

### CODE REVIEW

**Comment by gbggrant** [comment]
@yfarjoun is this ready for review?

File: src/main/java/picard/sam/CleanSam.java [hunk 1]
```diff
@@ -1,7 +1,7 @@
 /*
  * The MIT License
  *
- * Copyright (c) 2010 The Broad Institute
+ * Copyright (c) 2010 - 2016 The Broad Institute
```
Comment (gbggrant) [inline src/main/java/picard/sam/CleanSam.java:4]: What's so special about 2016 here?  Why not through 2021?

**Comment by gbggrant** [comment]
@yfarjoun is this ready to go?

File: src/main/java/picard/sam/CleanSam.java (hunk 1 above)
Comment (yfarjoun) [reply src/main/java/picard/sam/CleanSam.java:4]: goes to show when the code was written...I guess.

**Comment by yfarjoun** [comment]
I think it's waiting for a review....

**Comment by yfarjoun** [comment]
unless the only comment is the 2016 copyright notice.

File: src/main/java/picard/fastq/SamToBfqWriter.java [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+     *                       from throughout the file (null for all).
+     * @param chunk          the maximum number of records that should be written to any one file
+     * @param pairedReads    whether these reads are from  a paired-end run
+     * @param namePrefix     The string to be stripped off the read name
+     *                       before writing to the bfq file. May be null, in which case
+     *                       the name will not be trimmed.
+     * @param includeNonPfReads whether to include non pf-reads
+     * @param clipAdapters    whether to replace adapters as marked with XT:i clipping position attribute
+     */
+    public SamToBfqWriter(final File bamFile, final File referenceSequence, final String outputPrefix, final Integer total,
```
Comment (gbggrant) [inline src/main/java/picard/fastq/SamToBfqWriter.java:91]: Should this be `File samFile`? (and elsewhere)

File: src/main/java/picard/fastq/SamToBfqWriter.java [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+                }
+                if (chunk > 0 && wrote % chunk == 0) {
+                    initializeNextBfqFiles(fileIndex++);
+                }
+            }
+        }
+    }
+
+    /**
+     * Closes any the open bfq file(s), if any, and opens the new one(s)
```
Comment (gbggrant) [inline src/main/java/picard/fastq/SamToBfqWriter.java:262]: ??  
```suggestion
     * Closes any open bfq file(s), and opens the new one(s)
```

File: src/main/java/picard/fastq/SamToBfqWriter.java [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+    /**
+     * Count the number of records in the bamFile that could potentially be written
+     *
+     * @return  the number of records in the Bam file
+     */
+    private int countWritableRecords() {
+        int count = 0;
+
+        final SamReader reader = SamReaderFactory.makeDefault().referenceSequence(referenceSequence).open(bamFile);
+        if(!reader.getFileHeader().getSortOrder().equals(SAMFileHeader.SortOrder.queryname)) {
```
Comment (gbggrant) [inline src/main/java/picard/fastq/SamToBfqWriter.java:391]:
```suggestion
        if (!reader.getFileHeader().getSortOrder().equals(SAMFileHeader.SortOrder.queryname)) {
```

File: src/main/java/picard/sam/CompareSAMs.java [hunk 5]
```diff
@@ -53,8 +53,8 @@
         programGroup = DiagnosticsAndQCProgramGroup.class)
 @DocumentedFeature
 public class CompareSAMs extends CommandLineProgram {
-    static final String USAGE_SUMMARY = "Compare two input \".sam\" or \".bam\" files.  ";
-    static final String USAGE_DETAILS = "This tool initially compares the headers of SAM or BAM files. " +
+    static final String USAGE_SUMMARY = "Compare two input SAM/BAM/CRAM files.  ";
+    static final String USAGE_DETAILS = "This tool initially compares the headers input files. " +
```
Comment (gbggrant) [inline src/main/java/picard/sam/CompareSAMs.java:57]:
```suggestion
    static final String USAGE_DETAILS = "This tool initially compares the headers of the input files. " +
```
??

File: src/main/java/picard/sam/SplitSamByNumberOfReads.java [hunk 6]
```diff
@@ -131,14 +134,16 @@ protected int doWork() {
         int fileIndex = 1;
 
         Function<Integer, SAMFileWriter> createWriter = (index) ->
-                writerFactory.makeSAMOrBAMWriter(header, true, new File(OUTPUT, OUT_PREFIX + "_" + String.format("%04d", index) + BamFileIoUtils.BAM_FILE_EXTENSION));
+                writerFactory.makeSAMOrBAMWriter(header, true,
+                        new File(OUTPUT, String.format("%s_%04d.%s", OUT_PREFIX,index,extension)));
```
Comment (gbggrant) [inline src/main/java/picard/sam/SplitSamByNumberOfReads.java:138]: spacing.
```suggestion
                        new File(OUTPUT, String.format("%s_%04d.%s", OUT_PREFIX, index, extension)));
```

**Review by gbggrant — APPROVED** [review]
@yfarjoun looks good to me.  A couple of small comments.
And tests are failing.

**Comment by gbggrant** [comment]
Hi @yfarjoun are you planning on working further on this one?  We could take it over - let us know.

**Comment by yfarjoun** [comment]
I might be able to work on this...let me see.

**Comment by yfarjoun** [comment]
Yeah, I'll try to get back to this by the end of next week if that's fast
enough.

Y.

**Comment by gbggrant** [comment]
Sure - that would be great.  We'll try to give it a timely re-review if needed.

File: src/main/java/picard/fastq/SamToBfqWriter.java (hunk 2 above)
Comment (yfarjoun) [reply src/main/java/picard/fastq/SamToBfqWriter.java:91]: sure. elsewhere, in this file? or elsewhere in other files?

**Comment by yfarjoun** [comment]
@gbggrant this is done, but the tests are broken....any idea whats going on?

**Comment by cmnbroad** [comment]
@yfarjoun @gbggrant I'm guessing this branch has tests from a very old branch that were written before the CRAM update in 2020. The "no reference supplied" error now manifests as an `IllegalArgumentException`, not a `CRAMException`, which is only for internal CRAM errors.

**Comment by yfarjoun** [comment]
@cmnbroad I just rebased, and now the tests aren't even running.... :-( I dont' understand what's going on....

**Comment by cmnbroad** [comment]
@yfarjoun - Are you pushing to the correct remote (your fork instead of to the picard repo) ? This branch for this PR is in the main repo.

**Comment by yfarjoun** [comment]
@cmnbroad I rebased and pushed an entirely different branch.... 😊 

all good now.

**Comment by yfarjoun** [comment]
@cmnbroad Could you have a look here? I have one failing test and it has to do with cram-internals sanity checking (the checksum unexpectedly comes out to be 0 and run in Piacrds' (ReorderSam (not SortSam)...other programs do fine, and Samtools doesn't complain about file being corrupt or somesuch...

cheers! Yossi.

**Comment by cmnbroad** [comment]
@yfarjoun I don't see any failures like you're describing, but maybe I'm missing them ? The travis failures all appear to be because of (legitimate ?) command line parsing failures. If you point me to one I can take a closer look.

**Comment by yfarjoun** [comment]
hmmm. I'm getting different errors locally....will debug further. thanks.

**Comment by yfarjoun** [comment]
OK. @cmnbroad I fixed the other issues...only one remaining is the one I wanted your help on.....

**Comment by yfarjoun** [comment]
@cmnbroad:
```
Gradle suite > Gradle test > picard.sam.CramCompatibilityTest > testShouldWriteCRAMWhenCRAMWithReference[0](picard.sam.ReorderSam, SEQUENCE_DICTIONARY=testdata/picard/sam/test_cram_file.ref.fa, testdata/picard/sam/test_cram_file_coordinate_sorted.cram, testdata/picard/sam/test_cram_file.ref.fa) FAILED
    java.lang.RuntimeException: Block CRC32 mismatch, actual: 0000 expected: c622f71d
        at htsjdk.samtools.cram.structure.block.Block.read(Block.java:288)
        at htsjdk.samtools.cram.structure.CompressionHeader.<init>(CompressionHeader.java:90)
        at htsjdk.samtools.cram.structure.Container.<init>(Container.java:150)
        at htsjdk.samtools.cram.build.CramContainerIterator.containerFromStream(CramContainerIterator.java:48)
        at htsjdk.samtools.cram.build.CramContainerIterator.readNextContainer(CramContainerIterator.java:30)
        at htsjdk.samtools.cram.build.CramContainerIterator.hasNext(CramContainerIterator.java:58)
        at htsjdk.samtools.CRAMIterator.nextContainer(CRAMIterator.java:93)
        at htsjdk.samtools.CRAMIterator.hasNext(CRAMIterator.java:204)
        at htsjdk.samtools.CRAMIterator.advanceToAlignmentInContainer(CRAMIterator.java:170)
        at htsjdk.samtools.CRAMFileReader.queryUnmapped(CRAMFileReader.java:380)
        at htsjdk.samtools.SamReader$PrimitiveSamReaderToSamReaderAdapter.queryUnmapped(SamReader.java:560)
        at picard.sam.ReorderSam.doWork(ReorderSam.java:158)
        at picard.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:308)
        at picard.sam.CramCompatibilityTest.launchProgram(CramCompatibilityTest.java:265)
        at picard.sam.CramCompatibilityTest.testShouldWriteCRAMWhenCRAMWithReference(CramCompatibilityTest.java:112)
```

**Comment by cmnbroad** [comment]
@yfarjoun The problem is that the index file `testdata/picard/sam/test_cram_file_coordinate_sorted.cram.bai` doesn't correspond to the `testdata/picard/sam/test_cram_file_coordinate_sorted.cram`. The cram file itself is fine, but it looks like it was probably generated by samtools (I can see from the encoding strategy used that it definitely wasn't created by htsjdk), but AFAIK only htsjdk generates bai. I think you mixed up the test files and have a samtools cram with an htsjdk bai. The test passes if you remove the .bai and replace it with the .crai from `samtools index`, or if you regenerate the cram and .bai using GATK PrintReads.

**Comment by yfarjoun** [comment]
Thanks @cmnbroad. Is this a known issue (that a htsjdk cram cannot be indexed with GATK)? we should either fix it, or make sure that htsjdk refuses to index such files....no?

**Comment by yfarjoun** [comment]
@gbggrant I fixed the failing tests and rebased...I think this is ready to merge....final scan?

**Comment by cmnbroad** [comment]
@yfarjoun I'm not sure how you made that index file, but I'm not aware of any issue that needs to be fixed here. The only requirement is that the index you use was created for that exact cram file. It doesn't/shouldn't matter what software created the file, or the index, as long as they're paired. That .bai file was not made from that CRAM file, so they weren't paired - but I think you just had the wrong file. On the other hand, if you did run some GATK command that produced an index that doesn't work, then let me know how you did it, since I'd definitely want to look into it.

**Comment by yfarjoun** [comment]
I guess that's my misunderstanding @cmnbroad. I thought you said that it was because I made the index with GATK as opposed to with samtools....

File: src/test/java/picard/sam/CramCompatibilityTest.java [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+            assertCRAM(outputFile);
+        } else {
+            final File tmpDir = IOUtil.createTempDir(outputDir.getAbsolutePath(), program);
+            launchProgram(program, cramFile, tmpDir.getAbsolutePath(), parameters, null);
+            assertCRAMs(tmpDir);
+        }
+    }
+
+    private File createTempCram(String name) throws IOException {
+        return createTempFile(name, ".cram");
```
Comment (gbggrant) [inline src/test/java/picard/sam/CramCompatibilityTest.java:224]: Should this have a deleteOnExit()?

**Review by gbggrant — APPROVED** [review]
@yfarjoun looks good.  One minor comment.
So, can you push to master anymore?

File: src/test/java/picard/sam/CramCompatibilityTest.java (hunk 7 above)
Comment (yfarjoun) [reply src/test/java/picard/sam/CramCompatibilityTest.java:224]: isn't it 2 lines below?

Comment (gbggrant) [reply src/test/java/picard/sam/CramCompatibilityTest.java:224]: Yes.

**Comment by yfarjoun** [comment]
WOOT! 🎉

**Pull request merged**